### DISCUSSION TITLE
field array - validation on change

### DISCUSSION THREAD

**Discussion by n-margalit** [react-hook-form/react-hook-form#7550] [Q&A] [answered]
I am trying to use field array + yup validations
I set the form mode : "onChange" so all fields validations triggered on change
but field array validations are triggered only after submit
can you please help?
[here is a code sandbox](https://codesandbox.io/s/youthful-keldysh-6llkd?file=/src/Children.tsx)
TIA

**Comment by Moshyfawn** [comment]
Your `resolver.ts` is empty, so the app crashes: `_resolver.useYupValidationResolver is not a function`.
Did you mean to use #225 `useYupValidationResolver`?

**Reply by n-margalit** [reply]
fixed

**Comment by Moshyfawn** [comment, marked as answer]
Alright, so there're a couple of things that has lead to the issue described.
1. Doing `errors[props.name]` won't access the error message you need, since `props.name` is usually a hodgepodge of bracket and dot notations due to the nature of `useFieldArray` data structure.
2. The 1st point leads to this one: you should consider using dot notation to assign the `<TextField />` name; `'children.${index}.name'` instead of `'children[${index}].name' `. It's gonna allow you to actually access those dynamic field names by doing something like [this path resolver](https://stackoverflow.com/a/45322101/10414512). Additionally, you can see, that your `isHold` & `isBIL` checkboxes aren't checked by default after appending a new field, but your default `append` properties set as `{ isHold: true, isBIL: true }`. Converting names to dot notation fixes the issue; the checkboxes are checked by default after `append`.
3. I'm not sure what's the issue with the custom yup resolver `useYupValidationResolver` you're using, but switching to use the [@hookform/resovlers package](https://github.com/react-hook-form/resolvers) with the [yupResolver](https://github.com/react-hook-form/resolvers#yup) fixes the "not validating `onChange`" issue. So, I'd advise to just use this package instead of a custom solution. 

See this modified [CSB](https://codesandbox.io/s/rhf-field-array-validation-forked-7u7eh?file=/src/TextField.tsx) for a working 'onChange' validation example. Take notes on:
- how the `yupResolver` is used in `App.js`
- The `<TextField />`'s names with dot notation in `Children.tsx` `name={`children.${index}.name`}`
- A custom resolve function to access those dynamic `<TextField />` names switched to dot notation like `resolve(props.name, errors)?.message`

**Reply by Moshyfawn** [reply]
Also, for `append` method you might want to consider setting default values for all properties, not only for checkboxes; `null` or an empty string for `<TextField />`'s should be fine.

**Reply by n-margalit** [reply]
Thanks alot!
works fine now

**Reply by Moshyfawn** [reply]
No problem!

**Comment by Moshyfawn** [comment]
One more thing, @n-margalit. Just forgot to mention that you should use `field.id` instead of field array index.[ Refer to the Rules section of RHF docs on `useFieldArray`](https://react-hook-form.com/api/usefieldarray/):
> The field.id **(and not index)** must be added as the **component key to prevent re-renders** breaking the fields:
